src/qibolab/_core/instruments/emulator/hamiltonians.py

```diff
@@ -0,0 +1,112 @@
+from dataclasses import dataclass
+from functools import cached_property
+from typing import Literal, Optional
+
+import numpy as np
+from pydantic import Field
+
+from ...components import Config, IqConfig
+from ...pulses import Delay, Pulse
+from .operators import L1, L2, QUBIT_DRIVE, SIGMAZ
+from .utils import HZ_TO_GHZ
+
+
+class Qubit(Config):
+    """Hamiltonian parameters for single qubit."""
+
+    frequency: float = 0
+    """Qubit frequency for 0->1."""
+    t1: float = 0
+    """Relaxation time."""
+    t2: float = 0
+    """Coherence time."""
+
+    @property
+    def operator(self):
+        """Time independent operator."""
+        return -np.pi * self.frequency * HZ_TO_GHZ * SIGMAZ
+
+    @property
+    def t_phi(self):
+        """T_phi computed from T1 and T2."""
+        return 1 / (1 / self.t2 - 1 / self.t1 / 2)
+
+    @property
+    def decoherence(self):
+        """Decoherence operator."""
+        assert self.t1 > 0 and self.t2 > 0
```

Comment:
If they are not positive, you would just get a `nan` out of the square root, and `nan` propagates.

Sometimes, if there is just a chance of getting some `nan`, better to allow for a few `nan` rather than halting the whole run.
```suggestion
```

Reply:
Ok, it seems that Python is not fully implementing the standard for floats. You can see the difference with NumPy
```py
>>> np.arange(3) / 0
array([nan, inf, inf])
```

In any case, even if we migrated everything to NumPy, QuTiP would still raise the error above, which is less clear than your assertion.

At this point, definitely better to keep the `assert`.